contracts/AelinPool.sol

```diff
@@ -385,6 +445,83 @@ contract AelinPool is AelinERC20, MinimalProxyFactory, IAelinPool {
+        uint256 _purchaseTokenAmount
+    ) external lock {
+        require(hasNftList, "pool does not have an NFT list");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+        require(!nftWalletUsedForPurchase[msg.sender], "wallet already used");
+
+        NftData memory tmpNftData = nftCollectionDetails[_collectionAddress];
+        require(tmpNftData.collectionAddress == _collectionAddress, "collection not in the pool");
+
+        if (tmpNftData.purchaseAmountPerToken) {
+            require(
+                _purchaseTokenAmount <= tmpNftData.purchaseAmount * _tokenIds.length,
+                "purchase amount per token should be less than the allocation"
+            );
+        }
+
+        if (!tmpNftData.purchaseAmountPerToken && tmpNftData.purchaseAmount > 0) {
+            require(
+                nftAllowList[msg.sender] + _purchaseTokenAmount <= tmpNftData.purchaseAmount,
+                "purchase amount should be less than the allocation"
+            );
+            nftAllowList[msg.sender] += _purchaseTokenAmount;
+        }
+
+        if (NftCheck.supports721(_collectionAddress)) _purchasePoolWith721(_collectionAddress, _tokenIds);
+        if (NftCheck.supports1155(_collectionAddress)) _purchasePoolWith1155(_collectionAddress, _tokenIds, tmpNftData);
+        if (_collectionAddress == CRYPTO_PUNKS) _purchasePoolWithPunks(_collectionAddress, _tokenIds);
+
+        nftWalletUsedForPurchase[msg.sender] = true;
+        IERC20(purchaseToken).safeTransferFrom(msg.sender, address(this), _purchaseTokenAmount);
+        _mint(msg.sender, _purchaseTokenAmount);
+        emit PurchasePoolToken(msg.sender, _purchaseTokenAmount);
+    }
+
+    function _purchasePoolWith721(address _collectionAddress, uint256[] calldata _tokenIds) internal {
+        for (uint256 i = 0; i < _tokenIds.length; i++) {
+            require(IERC721(_collectionAddress).ownerOf(_tokenIds[i]) == msg.sender, "has to be the owner of the token");
+            require(!nftId[_collectionAddress][_tokenIds[i]], "tokenId already used");
+            nftId[_collectionAddress][_tokenIds[i]] = true;
+        }
+    }
+
+    function _purchasePoolWith1155(
+        address _collectionAddress,
+        uint256[] calldata _tokenIds,
+        NftData memory tmpNftData
+    ) internal view {
+        for (uint256 i = 0; i < _tokenIds.length; i++) {
+            require(nftId[_collectionAddress][_tokenIds[i]], "tokenId not in the pool");
+            require(
+                IERC1155(_collectionAddress).balanceOf(msg.sender, _tokenIds[i]) >= tmpNftData.minTokensEligible,
+                "has to be the owner of the token"
+            );
+        }
+    }
+
+    function _purchasePoolWithPunks(address _punksAddress, uint256[] calldata _tokenIds) internal {
+        for (uint256 i = 0; i < _tokenIds.length; i++) {
+            require(ICryptoPunks(_punksAddress).punkIndexToAddress(_tokenIds[i]) == msg.sender, "not the owner");
```

Comment:
i tried including this in `_purchasePoolWith721()` but it reverted back the require statement and didn't proceed with the function even when there was an OR. 

`require(ICryptoPunks(address).punkIndexToAddress(id) == msg.sender || IERC721(address).ownerOf(id) == msg.sender);`

that's why i added a new function just for punks

Reply:
I think this is a good function to separate. nice